Fixes and changes related to silencing HPUX build warnings

demerphq opened the pull request:
This is a collection of fixes that came up while trying to get things to build properly on HPUX Itanium. We were producing hundreds of warnings of different types, some actually valid warnings and issues, other cruft from HPUX being super picky and not understanding some of the tricks we use. 

In regard to false positive warnings, two notable examples are our OO in C type logic that we do with casting types with similar layouts from one type to another makes the HPUX compiler produce loads of "possible misalignment" warnings.  Since we test thoroughly and Itanium would segfault if we actually did use misaligned pointers these warnings can be silenced. Another case that produces literally hundreds of warnings is the memory wrap logic which produces a warning about a constant comparisons always being false. We hard silence these warning categories.

Another case is that the XS generated code from ParseXS.pm can produce code that is unreachable, so we silence those warnings on hpux as well.

Other examples are legitimate, for instance using && against a constant where we should be using &.

Another case is that the HPUX compiler whines about assignment in a while conditional, so I rewrote the code to not do that.

This includes some changes to the regnode structures used by the regex engine which I did as an initial response to the misalignment problem. I kept these changes as IMO they actually improve the code by making it more clear what parts of the regnode might have multiple meanings, and by ensuring all access to the contents of the regnode srtuctures is via macros which know the correct struct layout. This also means some of the declaration code can be refactored to share a common set of named unions.

There is a bugfix to cmp_version.t to deal with older git's that don't support taggerdate:unix.  

With all of these patch @Tux's HPUX box passes tests under use64bitall. (It fails GDBM tests when built without the use64bitall).

demerphq committed:
perl.h - silence certain warnings on HPUX globally.

There are two warnings classes which account for a very large number of
the warnings produced when building on HPUX Itanium. We know the cause
of these warnings and we are ok with ignoring them.

One set comes from our memory wrap checks, where we end up doing a
comparison against constants in certain conditions. See the comments in
handy.h line 2723 related to PERL_MALLOC_WRAP.

The other set comes from our common "trick" of doing OO in C code with
casting. This is the foundation of how we manage SV types and how we
manage regular expression ops (regops).

If this logic really was a problem then we would have lots of test
failures and segfaults, and we do not, so we can silence them.

demerphq committed:
regcomp.h - use different struct member names for U8 vs U32 str_len

It is confusing to have two different members, at different struct
offsets and with different sizes, with the same name. So rename them
so they have different names that include their size so it is obvious
what is going on.

demerphq committed:
regcomp.h - use a common union for head and args across all regnodes.

This helps with HPUX builds where we need to ensure everything
is aligned the same (on 32 bit boundaries). It also strongly
encourages everything to use the accessor macros and not access
the members directly.

By using a union for the variadic fields we make it more obvious
that some regops use the field in different ways. This patch
also converts all the arg unions into a standardized union with
standardized member names.

demerphq committed:
peep.c - ensure deferred_queue is initialized before use

HPUX was warning that it can be used before it is initialized.
There is no cost to nulling it out early so do so.

demerphq committed:
op.c - deal with not-reached statement warning

The return statement could never be reached as the while loop
is while(1) and there is no break statement inside of it. This
patch replaces the 'return arg;' with a NOT_REACHED; assertion,
just in case someone does add a break later on without thinking.

demerphq committed:
pp_sys.c - suppress warning about comparison to unsigned value

demerphq committed:
regcomp.c - simplify logic to eliminate unreached code

We always goto out of this if, so eliminate the unreached code
and also eliminate the useless else block.

demerphq committed:
regexec.c - mark the bottom of a sub as never reached

We used to have a return statement here that picky compilers
would complain about.

demerphq committed:
sv.c - don't cast SVTYPEMASK to svtype, use SvIS_FREED() instead.

svtype is an enum with 18 values. SVTYPEMASK is 31. A picky compiler
(like on HPUX) will complain that casting 31 to a svtype is an error.

We have SvIS_FREED() to do this properly anyway.

demerphq committed:
Porting/cmpVERSION.pl - deal with old git clients better in tag date parsing

The HPUX box we smoke on does not have a modern git and chokes
on taggerdate:unix, so we use taggerdata:iso8601 instead, and
then parse the date with Time::Local.

demerphq committed:
t/op/sprintf2 - TODO some failing tests for %g.

On HPUX none of the usual methods for doing high precision %g seem to
work:

    Checking for an efficient way to convert floats to strings.
    Trying sprintf...
    sprintf() found.
    sprintf length mismatch:  Expected 55, got 38
    ...But sprintf didn't work as I expected.
    Trying gconvert...
    gconvert NOT found.
    Trying gcvt...
    gcvt() found.
    gcvt length mismatch:  Expected 55, got 38
    ...But gcvt didn't work as I expected.
    *** WHOA THERE!!! ***
    None of ( sprintf gconvert gcvt)  seemed to work properly.  I'll use sprintf.

So we can safely TODO these tests for now.

See: https://github.com/Perl/perl5/issues/20953#issuecomment-1478744988
and: https://github.com/Perl/perl5/issues/20953#issuecomment-1483814118

Fixes: #20953
And also some issues in: #20959

demerphq committed:
ExtUtils-ParseXS - silence warnings about unreached code in generated XS code

On HPUX we get a lot of:

"APItest.c", line 4004: warning #2111-D: statement is unreachable: PUTBACK

because people have used something like XSRETURN(1); and we generate
code that does PUTBACK; return; Adding a pragma to silence it makes the builds
happier;

demerphq committed:
regcomp.c - eliminate assignment in while condition

This gives covertiy heartburn, and also makes the HPUX compiler
whine. Reworking it so that it is a while 1 and the condition determines
if we break out of the loop resolves the heartburn and makes for more
readable code.

demerphq committed:
numeric.c - change logical && to bitwise & in nantype test

This appears to have been broken since 2015, found by HPUX compiler warnings:

    "numeric.c", line 927: warning #4277-D: logical AND with a constant, do you
          mean to use '&'?
                          !(nantype && IS_NUMBER_IN_UV)) {

demerphq committed:
Devel-PPPort - deal with signed klen in check_HeUTF8

Some of the HV logic uses a negative key length to indicate utf8
keys, and this logic was using an unsigned keylength, which obviously
does not work

demerphq committed:
perlio.c - double cast PerlIO_lockcnt()

This avoids various warnings about alignment on HPUX.

demerphq committed:
util.c - assign HS_GETXSVERLEN() to var before assert

This makes it easier to see the value with gdb, and also avoids an
assignment in the if condition body.

demerphq committed:
util.c - avoid warning about truncated pointer value

This code is part of the fallback rng initialization, it should
only be used in odd circumstances, and it is intended to be "random"ish,
so the logic change should not be a problem.

This more or less reverts commit 4b4c4ab35e70f005167c863eae6d2003492c393c,
although the S_ptr_hash() now lives in util.c instead of hv.c.

Turns out having a decent pointer hash /is/ useful.

demerphq committed:
intrpvar.h - silence deprcation warning on HPUX

We were already silencing it on GCC, so lets add HPUX to the party.
Unfortunately I have not figured out a way to have a nice neat define
for it like we have for GCC or MSVC. :-(

Silences this:
    "intrpvar.h", line 941: warning #3215-D: function "Perl_sv_nounlocking"
              (declared at line 5934 of "proto.h") was declared deprecated
      PERLVARI(I, unlockhook,       share_proc_t, PERL_UNLOCK_HOOK)
      ^

We probably should add the same for MSVC as well.

demerphq committed:
intrpvar - ignore deprecated warning on MSVC as well

demerphq committed:
dist/threads - use pure pragma form of CLANG_DIAG_IGNORE in threads.xs

Otherwise on HPUX we get warnings

    "threads.xs", line 1057: warning #2111-D: statement is unreachable
          CLANG_DIAG_IGNORE_STMT(-Wthread-safety);

demerphq committed:
threads-shared - use pure pragma version of CLANG_DIAG_IGNORE

Otherwise HPUX complains about unreachable statements.

"shared.xs", line 707: warning #2111-D: statement is unreachable
        CLANG_DIAG_RESTORE_STMT;
        ^

This version should do the same thing, the only difference being
it doesnt allow a semicolon on the macro call.

mauke commented on peep.c:
```diff
@@ -2690,7 +2690,7 @@ Perl_rpeep(pTHX_ OP *o)
     OP* oldop = NULL;
     OP* oldoldop = NULL;
-    OP** defer_queue[MAX_DEFERRED]; /* small queue of deferred branches */
+    OP** defer_queue[MAX_DEFERRED] = { NULL, NULL, NULL, NULL }; /* small queue of deferred branches */
```
This could be shortened to
```c
OP** defer_queue[MAX_DEFERRED] = { NULL };
```
This decouples the initializer from the value of `MAX_DEFERRED`.

demerphq replied:
Ack. Thanks.

demerphq replied:
Fixed with the most recent push.

demerphq committed:
perl.c - show HAS_LONG_DOUBLE and HAS_STRTOLD as well as USE_LONG_DOUBLE

Makes it easier to debug long double issues.

demerphq committed:
hints/hpux - only run d_strtold define override check on version < 11.23

I added a version check around the logic as the comment already says
that "In HP-UXes prior to 11.23 strtold() returned a HP-UX specific
union called long_double, not a C99 long double."

The logic did not check for the version mentioned in the comment, it
simply greped for something that is no longer present in the file:

    grep 'double strtold.const' /usr/include/stdlib.h

outputs nothing on HPUX version B.11.31.

But

    grep "double strtold" /usr/include/stdlib.h

outputs:

    extern long_double strtold __((const char * __restrict, char ** __restrict));
    extern long double strtold __((const char * __restrict, char ** __restrict));

on HPUX version B.11.31.

demerphq committed:
t/porting/globvar.t - fix for HPUX

On HPUX `nm globals.o` produces output like this (without the indent):

    [5]    |                  2420|       2|OBJT |GLOB |0| .rodata|PL_Yes

So change the $define qr// to accommodate it.

We also have to TODO some of the tests, as HPUX seems to export
everything.

demerphq commented:
@Tux: with this patch sequence applied all tests pass on HPUX with use64bitall and uselongdouble.

IMO we should merge this sequence ASAP so that HPUX builds clean and passes test on 5.38. 

Reviews appreciated.

demerphq committed:
Porting/maintainers.pl - remove Net-Ping customized files

Net-Ping is in dist/ which means we are upstream, so there should
not be any customized files.

demerphq committed:
dist/Net-Ping - remove HPUX from the exclusions list in 450_service.t

We pass these tests just fine on HPUX 11.31.

demerphq commented:
Id like to merge this soon, so unless there are objections registered ill merge it in the next day or so.

tonycoz commented on op.c:
```diff
@@ -2469,8 +2469,8 @@ Perl_scalarvoid(pTHX_ OP *arg)
     }
-
-    return arg;
+    /* we should never get here */
```
Given the code at the end of the loop I don't think it's possible for that code to be reached.

Does simply changing this to `NOT_REACHED;` prevent the warning?

demerphq replied:
Not sure what you mean by "changing this". I tried adding a NOT_REACHED before the `return arg;` but it still threw warnings. The HPUX compiler doesn't seem to recognize various conventions that gcc does.

tonycoz replied:
Sorry, I meant `NOT_REACHED;` instead of the return.

Though I'm surprised we don't see warnings for the missing return with the change as it is.

demerphq replied:
dmq: i think i prefer while(1) + NOT_REACHED.
TonyC: I'm fine with that

tonycoz commented on pp_sys.c:
```diff
@@ -2989,7 +2989,13 @@ PP(pp_stat)
             Stat_t s;
             CLANG_DIAG_IGNORE_STMT(-Wtautological-compare);
             GCC_DIAG_IGNORE_STMT(-Wtype-limits);
+#if defined(__HP_cc) || defined(__HP_aCC)
+#pragma diag_suppress 2186
+#endif
             neg = PL_statcache.st_ino < 0;
+#if defined(__HP_cc) || defined(__HP_aCC)
+#pragma diag_default 2186
+#endif
             GCC_DIAG_RESTORE_STMT;
             CLANG_DIAG_RESTORE_STMT;
```
We can avoid all of this - config.h defines `ST_INO_SIGN` which has a negative value if `st_ino` is signed.

Especially since "fsblkcnt_t, fsfilcnt_t, and ino_t shall be defined as unsigned integer types" per POSIX.

demerphq replied:
Ok. Ill change it to use that instead. Are you thinking i should remove the DIAG_IGNORE statements as well?  I have to admit i tend to think that it would be better to merge this with the HPUX diag ignores, then afterward do a follow up that removes all of them. What do you think?

demerphq replied:
I have to admit I am a /little/ hesitant about this one.  The comment just above this says:
```
           /*
             * We try to represent st_ino as a native IV or UV where
             * possible, but fall back to a decimal string where
             * necessary.  The code to generate these decimal strings
             * is quite obtuse, because (a) we're portable to non-POSIX
             * platforms where st_ino might be signed; (b) we didn't
             * necessarily detect at Configure time whether st_ino is
             * signed; (c) we're portable to non-POSIX platforms where
             * ino_t isn't defined, so have no name for the type of
             * st_ino; and (d) sprintf() doesn't necessarily support
             * integers as large as st_ino.
             */
```

Which seems to suggest that we don't reliably detect that st_ino is signed. I would rather we don't change this as part of this PR, and do a separate PR where we try to change it.

tonycoz replied:
Ok

tonycoz commented on hv.c:
```diff
@@ -480,7 +480,7 @@ Perl_hv_common(pTHX_ HV *hv, SV *keysv, const char *key, STRLEN klen,
     if (!hv)
         return NULL;
-    if (SvTYPE(hv) == (svtype)SVTYPEMASK)
+    if ((U8)SvTYPE(hv) == (U8)SVTYPEMASK)
```
These should all use `SvIS_FREED(sv)`.

(I started typing "Maybe we should have a macro...".)

demerphq replied:
Hmm.  I didnt know we had that one. But the code would be a little different:

```
sv.h:#define SvIS_FREED(sv)     UNLIKELY(((sv)->sv_flags == SVTYPEMASK))
sv.h:#define SvTYPE(sv) ((svtype)((sv)->sv_flags & SVTYPEMASK))
```

Are you confident that would be safe?

tonycoz commented on dist/threads-shared/lib/threads/shared.pm:
```diff
@@ -8,7 +8,7 @@ use Config;
 
 use Scalar::Util qw(reftype refaddr blessed);
 
-our $VERSION = '1.67'; # Please update the pod, too.
```
"prgagma" in commit subject.

demerphq replied:
ack, thanks.

tonycoz commented on perl.h:
```diff
@@ -5723,6 +5723,12 @@ EXTCONST char PL_bincompat_options[] =
+#  ifdef HAS_LONG_DOUBLE
+                             " HAS_LONG_DOUBLE"
+#  endif
+#  ifdef HAS_STRTOLD
```
I think `HAS_STRTOLD` belongs in `non_bincompat_options[]` in perl.c, I don't see how it has an effect on binary compatibility.

That said, some of the existing `non_bincompat_options` probably belong in `PL_bincompat_options`.

demerphq replied:
Ok, ill take a look.
